text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+For more details on Intel SGX remote attestation we refer to the vendor [documentation](https://software.intel.com/content/www/us/en/develop/topics/software-guard-extensions/attestation-services.html).
+
+Once the chaincode definition for an FPC chaincode has been approved by the consortium, a Chaincode Enclave is created by calling `peer lifecycle chaincode createenclave`. This command triggers the creation of an enclave and registers it at the FPC Registry.  The registration includes the attestation of the Chaincode Enclave.
+
+The FPC Registry stores all attestation reports, as signed by the TEE vendor. 
+The attestation reports include information on what chaincode is running; what specific TEE is in use, including version number and hash of the chaincode (MRENCLAVE), and including the public keys of the enclave. 
+Upon creation each chaincode enclave generates public/private key pairs for signing and for encryption. The public signing key also denotes the chaincode enclave identity.
+
+In order to support the endorsement model of Fabric, chaincode enclaves executing the same FPC chaincode need access to a set of shared secrets. In particular, these shared or common keys are used to encrypt and decrypt the transaction arguments and state.
+That is, there is also a public/private key pair and a symmetric state encryption key per FPC chaincode that are shared among all chaincode enclaves that run the same FPC chaincode using the key distribution protocol.
+
+The deployment is described in detail below and can be found in the [Full Detail Diagrams](#full-detail-diagrams) section.
+
+
+## FPC Transaction Flow
+
+To illustrate how the FPC architecture works and how it ensures robust end-to-end trust, we describe the process of an FPC transaction. This assumes that all of the above described elements are already in place. Detailed explanations of that process are included below in the [Reference-Level Explanation](#reference-level-explanation).
+
+* Step 1: Client Invocation of the FPC Chaincode
+
+	![Invoke](../images/fpc/high-level/FPC-Invoke.png)
+
+	The Client prepares the Invocation of an FPC Chaincode by first encrypting the arguments of the Chaincode Invocation using the public key specific to a particular Chaincode. This encryption happens completely transparently using our FPC Client SDK extension. This Transaction Proposal is then sent to the Endorsing Peer where a corresponding Chaincode Enclave resides. Depending on the Endorsement Policy the client may perform this step with one or more Endorsing Peers and their respective Chaincode Enclaves. (For simplicity we will continue describing the process for a single Endorsing Peer.) The Peer forwards the Transaction Proposal to its FPC Chaincode running inside the Chaincode Enclave. Inside the Enclave, the FPC Shim decrypts the Proposal and invokes the FPC Chaincode.
```

Comment:
> The Client prepares the Invocation of an FPC Chaincode by first encrypting the arguments of the Chaincode Invocation using the public key specific to a particular Chaincode.

I guess it needs to encrypt it several times for multiple peers, because the proposal response contains a hash of the proposal?

Reply:
The public key is not unique to an enclave but corresponds to the chaincode overall, i.e., all enclaves authorized for that chaincode share the  private key corresponding to the chaincode public key

Reply:
So where is the private key generated and how does it gets to the enclaves?

Reply:
It is created by the initial enclave itself with the public key attested as part of the enclave registration and the private key (together with the state encryption key) distributed to other authorized enclaves as part of the key distribution protocol.